query/src/storages/fuse/pruning/pruning_executor.rs

```diff
@@ -0,0 +1,182 @@
+use tracing::Instrument;
+
+use super::bloom_pruner;
+use crate::sessions::TableContext;
+use crate::storages::fuse::io::MetaReaders;
+use crate::storages::fuse::pruning::limiter;
+use crate::storages::fuse::pruning::range_pruner;
+
+pub struct BlockPruner {
+    table_snapshot: Arc<TableSnapshot>,
+}
+
+const FUTURE_BUFFER_SIZE: usize = 10;
+
+impl BlockPruner {
+    pub fn new(table_snapshot: Arc<TableSnapshot>) -> Self {
+        Self { table_snapshot }
+    }
+
+    // prune blocks by utilizing min_max index and bloom filter, according to the pushdowns
+    #[tracing::instrument(level = "debug", skip(self, schema, ctx), fields(ctx.id = ctx.get_id().as_str()))]
+    pub async fn prune(
+        &self,
+        ctx: &Arc<dyn TableContext>,
+        schema: DataSchemaRef,
+        push_down: &Option<Extras>,
+    ) -> Result<Vec<(usize, BlockMeta)>> {
+        let segment_locs = self.table_snapshot.segments.clone();
+
+        if segment_locs.is_empty() {
+            return Ok(vec![]);
+        };
+
+        // if there are ordering clause, ignore limit, even it has been pushed down
+        let limit = push_down
+            .as_ref()
+            .filter(|p| p.order_by.is_empty())
+            .and_then(|p| p.limit);
+
+        let filter_expression = push_down.as_ref().and_then(|extra| extra.filters.get(0));
+
+        // shortcut, just returns all the blocks
+        if limit.is_none() && filter_expression.is_none() {
+            return Self::all_the_blocks(segment_locs, ctx.as_ref()).await;
+        }
+
+        // 1. prepare pruners
+
+        // prepare the limiter. in case that limit is none, an unlimited limiter will be returned
+        let limiter = limiter::new_limiter(limit);
+
+        // prepare the range filter.
+        // if filter_expression is none, an dummy pruner will be returned, which prunes nothing
+        let range_filter_pruner =
+            range_pruner::new_range_filter_pruner(ctx, filter_expression, &schema)?;
+
+        // prepare the bloom filter, if filter_expression is none, an dummy pruner will be returned
+        let dal = ctx.get_storage_operator()?;
+        let bloom_filter_pruner =
+            bloom_pruner::new_bloom_filter_pruner(ctx, filter_expression, &schema, dal)?;
+
+        // 2. kick off
+        //
+        // To make the pruning process more parallel (not just concurrent), we explicitly spawn
+        // pruning tasks in (the multi-threaded) storage runtime.
+        //
+        // NOTE:
+        // A. To simplify things, an optimistic way of error handling is taken: errors are handled
+        // at the "collect" phase. e.g. if anything goes wrong, we do not break the whole
+        // pruning task immediately, but only at the time that all tasks are done
+        //
+        // B. since limiter is working concurrently, we arrange some checks among the pruning,
+        //    to avoid heavy io operation vainly,
+        let storage_runtime = ctx.get_storage_runtime();
```

Comment:
Resource race for all users?

Reply:
as you suggested, a temp new dedicated runtime will be introduced (WIP, with other tweaks)